src/GraphQL/Types/TypeInformation.cs

```diff
@@ -0,0 +1,258 @@
+                    }
+                    if (type.Type.IsGenericType)
+                    {
+                        if (IsRecognizedListType(type.Type))
+                        {
+                            //unwrap type and mark as list
+                            IsList = true;
+                            ListIsNullable = type.Nullable != NullabilityState.NotNull;
+                            continue;
+                        }
+                    }
+                    if (type.Type == typeof(IEnumerable) || type.Type == typeof(ICollection))
+                    {
+                        //assume list of nullable object
+                        IsList = true;
+                        ListIsNullable = type.Nullable != NullabilityState.NotNull;
+                        break;
+                    }
+                }
+                //found match
+                IsNullable = type.Nullable != NullabilityState.NotNull;
+                Type = type.Type;
+                return;
+            }
+            //unknown type
+            IsNullable = true;
+            Type = typeof(object);
+        }
+
+        /// <summary>
+        /// Flattens a complex <see cref="NullabilityInfo"/> structure into a list of types and nullability flags.
+        /// <see cref="Nullable{T}"/> structs return their underlying type rather than <see cref="Nullable{T}"/>.
+        /// </summary>
+        private static List<(Type Type, NullabilityState Nullable)> Interpret(NullabilityInfo info, bool isInputProperty)
+        {
+            var list = new List<(Type, NullabilityState)>(info.GenericTypeArguments.Length + 1);
+            RecursiveLoop(info);
+            return list;
+
+            void RecursiveLoop(NullabilityInfo info)
+            {
+                if (info.Type.IsGenericType)
+                {
+                    var nullableType = Nullable.GetUnderlyingType(info.Type);
+                    if (nullableType != null)
+                    {
+                        list.Add((nullableType, NullabilityState.Nullable));
+                    }
+                    else
+                    {
+                        list.Add((info.Type, isInputProperty ? info.ReadState : info.WriteState));
+                    }
+                    foreach (var t in info.GenericTypeArguments)
+                    {
+                        RecursiveLoop(t);
+                    }
+                }
+                else if (info.ElementType != null)
+                {
+                    list.Add((info.Type, isInputProperty ? info.ReadState : info.WriteState));
+                    RecursiveLoop(info.ElementType);
+                }
+                else
+                {
+                    list.Add((info.Type, isInputProperty ? info.ReadState : info.WriteState));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Applies <see cref="GraphQLAttribute"/> attributes for the specified member to this instance.
```

Comment:
Not only `GraphQLAttribute`, `RequiredAttribute` too. I am not quite understand the responsibility of this class. Is it designed to handle **only** nullability information? If so why `IsList` here. If no then maybe to absorb other attributes into to handle all of them in one place?
![изображение](https://user-images.githubusercontent.com/21261007/151662242-1f0083a7-ae23-4303-b988-be288b73ac54.png)

Reply:
I'm not sure exactly what class you are referring to, since you posted the question in the `TypeInformation` class, referenced the `GraphQLAttribute` class, and posted a screenshot of the `AutoRegisteringHelper` class.  But let me try to clarify:

## TypeInformation

For `TypeInformation`, the purpose is to read a .NET `Type` and deconstruct that into bits that can be used to construct a graph type.  Graph types separate list and nullability flags from the `Type` and the NRT annotations.  There is also distinct behavior depending on if it's used as an input or as an output. So as "inputs" to this class, we need at a minimum the `PropertyInfo` to be parsed and to know if it's being used as a `get;` property or as a `set;` property.  We see that in the constructor:

```cs
public TypeInformation(PropertyInfo propertyInfo, bool isInputProperty)
```

Then the output is only the graph type.  We see that here:

```cs
public virtual Type ConstructGraphType()
```

We also allow `GraphQLAttribute`s marked on the property to modify this type information.  That is applied within this method:

```cs
public virtual void ApplyAttributes()
```

That covers everything in this entire class (besides the transient properties).  Keep in mind that when the attributes are applied, it only executes one specific overload of `GraphQLAttribute` -- the one that modifies a `TypeInformation` class, not the one that modifies a `FieldType` class:

```cs
public virtual void Modify(TypeInformation typeInformation)
```

So as an example, the user may have written this:

```cs
[Name("Id")]
[Id]
public int PersonId { get; set; }
```

We understand this to be a field named "Id" which is an `NonNullGraphType<IdGraphType>`.  But this happens with two independent parts of code.  One attribute overrides  `Modify(TypeInformation)`:

```cs
public class IdAttribute : GraphQLAttribute
{
    public override void Modify(TypeInformation typeInformation)
        => typeInformation.GraphType = typeof(IdGraphType);
}
```

While the other overrides `Modify(FieldType, bool)`:

```cs
public class NameAttribute : GraphQLAttribute
{
    // ...
    public override void Modify(FieldType fieldType, bool isInputType)
        => fieldType.Name = Name;
}
```

It can allow for an attribute to emit complex behavior, such as a a `MyCustomIdAttribute` containing both behaviors, and they both would get applied (in different areas) to the same property.

Note that while it is possible for the selected graph type to be changed by the `Modify(FieldType, bool)` method (and in fact this is how the `OutputType` attribute works), this would be rather cumbersome to retain the nullability and/or list wrappers, and it would not have access to the original `PropertyInfo` instance or original `Type` of the property.

## GraphQLAttribute

The purpose of this class is to apply changes to the schema as it is being built -- either by `AutoRegisteringObjectGraphType`, `AutoRegisteringInputObjectGraphType`, or `SchemaBuilder`.

Depending on where the derived attribute is placed, it may override one or another, or multiple, of its overridable methods.  Each method affects a different part of the schema.

| Method | Description |
|-|-|
| `Modify(TypeConfig)` | Provides configuration to the `SchemaBuilder` for types |
| `Modify(FieldConfig)` | Provides configuration to the `SchemaBuilder` for fields |
| `Modify(IGraphType)` | Provides configuration to the auto-registering classes for types |
| `Modify(FieldType, bool)` | Provides configuration to the auto-registering classes for fields |
| `Modify(TypeInformation)` | Changes graphtype-inference behavior for auto-registering fields |

There is only one attribute that does not fit within the above -- the `[Ignore]` attribute.  It seemed to me that the attribute was designed to control how the auto-registering classes scanned for fields to be added, rather than modifying a method or field as it was being built.  If you think we should change it, I can add another overridable method to `GraphQLAttribute` such as `bool ShouldInclude(MemberInfo)` or some such thing.

## GraphQL attributes vs component model attributes

The design here is that types and fields are initialized with "defaults" consisting of the following:
- Name derived from class or property name
- Description derived from `[Description]` attribute or xml comments, if those global options are enabled
- Deprecation reason derived from `[Obsolete]` attribute, if the global option is enabled
- Default value derived from `[DefaultValue]` attribute, if the global option is enabled

Then `GraphQLAttribute` attributes are applied afterwards.  Of course we could add our own `DescriptionAttribute` which would derive from `GraphQLAttribute` and set the description on a type or field.  But I saw no need for that.

Likewise the `TypeInformation` constructor performs type parsing by default, and then the attributes are applied.

## Custom attribute examples

- `[Nullable]` - force a field to be nullable, overriding NRT annotations
- `[ListNullable]` - force the list wrapper to be nullable, overriding NRT annotations
- `[MemberDefault]` - for complex default values, set a default value based on the value of a static property
- `[Auto]` - force a graph type to be `AutoRegisteringObjectGraphType<T>` or similar
- `[Model]` - append the word "Model" (or "Input") to the end of the name as a convention

I have several additional uses for attributes that I will be presenting in additional PRs that will be really cool.

Reply:
>  If you think we should change it, I can add another overridable method to `GraphQLAttribute` such as bool ShouldInclude(MemberInfo) or some such thing.

70/30. I think that it's better to do so now then some time after since it allows to
1) dynamically ignore fields, for example, to inherit from base attribute and use `[EnvIgnore(Environment: "PROD")] 
```
public class EnvIgnoreAttribute: IgnoreAttribute
{
    public string Environment { get; set; }
    public override bool ShouldInclude(....) => Environment.Current != Environment;
}
```
2) have the same design for all attributes

> Of course we could add our own DescriptionAttribute which would derive from GraphQLAttribute and set the description on a type or field. But I saw no need for that.

50/50. I would like to have single understandable design for all "graphql-related" dattributes. The disadvantage of the use of system.componentmodel attributes  is the possible conflict of interest. I came across it several times. Not big deal but may be a problem. Also inheritance from the base `GraphQLAttribute` will simplify the discoverability of these attributes and their use as I think but I'm not sure 100%. Of course shifting to such new attributes for descriptions/deprecations/default values is a breaking change, therefore I am not sure whether it is worth it.

Reply:
Sure let's put ShouldInclude in GraphQlAttribute.  Can't do it just now, but a little later today.